File: 0446-arithmetic-slices-ii-subsequence/0446-arithmetic-slices-ii-subsequence.cpp
class Solution {
public:
    int numberOfArithmeticSlices(vector<int>& nums) {
        int res = 0;
        vector<unordered_map<long long, int>> diff(nums.size());
        for (int i = 0; i < nums.size(); i++) {
            for (int j = 0; j < i; j++) {
                long long d = (long long)nums[i] - nums[j];
                if (!diff[j].count(d)) {
                    diff[i][d]++;
                } 
                else {
                    res += diff[j][d];
                    diff[i][d] += diff[j][d] + 1;
                }
            }
        }
        return res;
    }
};

File: 0446-arithmetic-slices-ii-subsequence/0446-arithmetic-slices-ii-subsequence.java
class Solution {
    public int numberOfArithmeticSlices(int[] nums) {
        int n = nums.length;
        Map<Integer, Integer>[] dp = new HashMap[n];
        for (int i = 0; i < n; i++) {
            dp[i] = new HashMap<>();
        }
        int result = 0;
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                long diffL = (long) nums[i] - nums[j];
                if (diffL < Integer.MIN_VALUE || diffL > Integer.MAX_VALUE) {
                    continue;
                }
                int diff = (int) diffL;
                dp[i].put(diff, dp[i].getOrDefault(diff, 0) + 1);
                if (dp[j].containsKey(diff)) {
                    dp[i].put(diff, dp[i].get(diff) + dp[j].get(diff));
                    result += dp[j].get(diff);
                }
            }
        }
        return result;
    }
}